packages/flutter/lib/src/painting/image_provider.dart

```diff
@@ -868,6 +887,104 @@ class ExactAssetImage extends AssetBundleImageProvider {
   String toString() => '${objectRuntimeType(this, 'ExactAssetImage')}(name: "$keyName", scale: $scale, bundle: $bundle)';
 }
 
+/// The action a [DeferringImageProvider] should take currently.
+enum DeferringImageProviderAction {
+  /// Indicates that the image is no longer needed, and the
+  /// [DeferringImageProvider] should stop seeking to resolve the image.
+  ///
+  /// After returning this action, the [DeferringImageProvider] will not call
+  /// its [DeferringImageProvider.getNextAction] callback anymore.
+  cancel,
+
+  /// Indicates that the caller would like to defer the image and be called
+  /// again after the next [DeferringImageProvider.deferDuration] elapses (or
+  /// at the end of current task queue if the duration is [Duration.zero]).
+  defer,
+
+  /// Indicates that the [DeferringImageProvider] should attempt to resolve the
+  /// image.
+  ///
+  /// After returning this action, the [DeferringImageProvider] will not call
+  /// its [DeferringImageProvider.getNextAction] callback anymore.
+  resolve,
+}
+
+/// An [ImageProvider] that can delay resolving an image using its wrapped
+/// [imageProvider] until an arbitrary condition is met.
+///
+/// For example, the [Image] widget uses this provider to defer loading images
+/// if it is the child of a scrollable that is rapidly scrolling. It cancels
+/// resolving the image if its element is evicted from the tree before scrolling
+/// ends, and otherwise waits for scrolling to slow down before actually
+/// resolving the image. This prevents the framework from over-utilizing
+/// hardware resources for images that will never be shown on screen.
+///
+/// If the image is already in the cache, e.g. from [precacheImage] or because
+/// another provider has already resolved it, the [getNextAction] callback will
+/// never fire and the cached image is reused.
+@immutable
+class DeferringImageProvider<T> extends ImageProvider<T> {
+
+  /// Creates a new [DeferringImageProvider].
+  ///
+  /// All parameters must not be null.
+  const DeferringImageProvider({
+    @required this.imageProvider,
+    @required this.getNextAction,
+  }) : assert(imageProvider != null),
+       assert(getNextAction != null);
+
+  /// The wrapped [ImageProvider] to delegate loading the image and creating
+  /// a key for it.
+  ///
+  /// The image provider's load method may never be called if resolution gets
+  /// cancelled.
+  final ImageProvider<T> imageProvider;
+
+  /// A callback used by [resolve] to determine whether resolution should be
+  /// cancelled, deferred, or executed.
+  ///
```

Comment:
Could you document at what point this is called? E.g. when do you get to make this decision?